### ISSUE TITLE
Avoid using global init script for developers

### ISSUE THREAD

**Issue by lacasseio** [NationalSecurityAgency/ghidra#459]
**Is your feature request related to a problem? Please describe.**
Global init scripts, as [described in the dev setup guide](https://github.com/NationalSecurityAgency/ghidra/blob/master/DevGuide.md#setup-source-and-dependency-repositories), are applied to all Gradle project ran on the machine. This is bad practice.

**Describe the solution you'd like**
Instead, Gradle can download and unzip the files when required and set up a temporary project-wide flat repository. All the steps explained in the dev guide can be run within Gradle tasks when the dependency is required.

**Describe alternatives you've considered**
Use a standalone init script that you pass to the project [using `--init-script`](https://docs.gradle.org/current/userguide/init_scripts.html#sec:using_an_init_script).

**Additional context**
None.

**Comment by nsadeveloper789** [comment]
I like the idea of using `--init-script` in the dev guide. We made quick mention of it existing, but didn't appreciate all the implications of setting that up globally. I think that's an easy enough adjustment to make, and we could even place it in the repo, though not enabled by default.

As for the automated download-and-unpack steps, you make a good point. We've talked a little internally about it as we were preparing our open-source release. It was less engineering effort to write steps into the dev guide than to write up tasks. We tend also to avoid automatic downloads unless we make it clear to the user/developer that it's happening. I think this is easily doable while addressing that concern, though. (Probably naming the tasks "download..." should be sufficient). At the time, we didn't have #359, but it's technically possible to automate that step now, too.

That said, we're currently refactoring our build scripts to better address common per-subproject concerns. I think it's best we wait to automate the dev guide steps until those changes are completed and merged.

**Comment by lacasseio** [comment]
I'm open for a quick chat regarding best practice and improvement regarding the refactoring you are under taking. I'm one of the Gradle engineer working on the native support. I'm sure I can provide a couple good pointers.

**Comment by nsadeveloper789** [comment]
From what I understand, his work is already being reviewed, so probably too late to chat, unfortunately. Nevertheless, in case we're off the rails, I'll give you an overview: Currently, we use `allprojects {}` and `subprojects {}` to handle just about every concern. Granted, not all projects share those concerns, so we tend to have many blocks conditionally excluded based on the value of some variable set by each project (so we end up using afterEvaluate {} a lot, too), or by pattern matching the project's file-system path (eww). Check out the `assembleSourceCommon` task in `gradleScripts/distribution.gradle` if you want to see just one sample. He's moved those concerns out into scripts and avoided `all/subprojects`. Instead each project applies only the scripts for the concerns it needs. This has already cleaned things up considerably.

Certainly, a good long term plan would be to factor those into plugins. We'd experimented with plugins before, but we sorely missed all of Eclipse's assistance editing the build scripts. It was easier when we needed to grep only `*.gradle` files, so we opted for a script-only solution. I understand Gradle Kotlin has better assistance, but we haven't investigated that option, yet.

I'm also aware of some upcoming changes to the way Gradle builds natives. If I'm not mistaken, that's available now, and the way we're currently doing it will soon be deprecated in Gradle 6.0? The above refactoring work does not address this. If you have any pointers here, we'd greatly appreciate it, though.

**Comment by lacasseio** [comment]
Plugins would definitely be a great way to refactor the logic. I understand the concern about the Intellisense not working properly using Eclipse generator plugin. There is ensentially two ways to write the plugins, either using `buildSrc` or standalone project included as composite build. However, if I'm not mistaken, using Eclipse plugin named Buildship should help a lot with the integration. Alternatively, Intellij has great, if not the best, support for Gradle and editing Gradle code especially when using Kotlin.

The usual way to configure a project in Gradle is to abstract the convention inside a plugin and configure the plugin in the project. You can apply the plugin to multiple projects using `allprojects`/`subprojects` but sometimes the projects are different enough that it may cause more issue than applying the right plugin to the right projects.

As for the native support, we are phasing out the software model plugins but won't remove them until the most important features are ported to the new plugins. More information on this will soon be announced.

**Comment by nsadeveloper789** [comment]
We used the `buildSrc` method in our experiments. We had the plugin project loaded in Eclipse, and we used BuildShip for editing buildscripts, but it still lacked good content assistance. We would probably need to move to Gradle Kotlin before we experiment with plugins again. Unfortunately, this is one of those "if it ain't broke, don't fix it" sort of things.... I'd be nice to follow best practices (and make a nice set of Gradle plugins available for extension developers), but we're OK(-ish) with the way things are, now.

Thanks for the info about natives. It's good to know the rug's not coming out from under us, yet.
